### PR TITLE
Implement Faucet Interface

### PR DESCRIPTION
**Pull request by Sharqiewicz**
Screenshot 2024-07-01 at 10 23 26

## Implement Faucet Interface
✅ Setup SvelteKit app
✅ Implement faucet form
✅ Connect interface with Express server
✅ Document development hints in README.md

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -41,7 +41,7 @@ export class Config {
         process.env.RATE_LIMIT_NUMBER_OF_PROXIES || "1",
       ),
     };
-    this.substrateAccountSecret = process.env.SUBSTRATE_SECRET_PHRASE!;
+    this.substrateAccountSecret = "abde";
```
Comment (ebma) [inline src/config.ts]: Let's revert this change 😅

File: src/slack_service/slack.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@ export class SlackNotifier {
   constructor(config: Config) {
     this.config = config;
 
+    process.env.SLACK_WEB_HOOK_TOKEN = "abdc";
+
```
Comment (ebma) [inline src/slack_service/slack.ts]: Let's remove

File: src/server.ts [hunk 3]
```diff
@@ -5,6 +5,7 @@ import { ApiManager } from "./chain_service/api.js";
 import Faucet from "./chain_service/faucet.js";
 import { AddressError, RateError } from "./chain_service/faucetErrors.js";
 import { SlackNotifier } from "./slack_service/slack.js";
+import { handler } from "../interface/build/handler.js";
```
Comment (ebma) [inline src/server.ts]: What's the handler doing? I don't think we should import something from a `build` directory as that file might not always exist.

File: interface/.gitignore
Comment (ebma) [inline interface/.gitignore:1]: I think we should add `.js` and `.js.map` files to the .gitignore, as I doubt we need to check them into source control?

**Review by ebma — COMMENTED** [review]
Cool 👍  I left some comments and would still need to test this locally. 
Could you please add some notes to the description that briefly describe how the faucet service itself is coupled to the frontend? I'm particularly interested in this `handler` that we now use in `src/server.ts`.

File: src/server.ts (hunk 3 above)
Comment (Sharqiewicz) [reply src/server.ts]: I took the whole idea of importing handler.js here from the official Svelte documentation: https://kit.svelte.dev/docs/adapter-node#custom-server

If we want our svelte application to be served via Express - we can do it with handler.js imported from frontend build. I think importing from the build is not a problem, if we configure Netlify correctly to create the build frontend first, then handlers.js will always exist

File: interface/.gitignore
Comment (Sharqiewicz) [reply interface/.gitignore:1]: Very good catch - You're right, I'll change it ✅

**Comment by Sharqiewicz** [comment]
Faucet after adding the interface supports the same functionality as before. The interface is added to Express using handler with handler.js from the frontend build, as documented in the official documentation.

The handler function from SvelteKit serves multiple roles within the Express application:


- It handles all frontend routes defined within the SvelteKit application, allowing seamless navigation and client-side routing.
- It serves static assets - JavaScript, CSS, and images included in the SvelteKit build.

**Comment by ebma** [comment]
The project doesn't compile for me. 

When I run `yarn build` from the root directory I get
```
$ yarn build
error TS6059: File '/Users/marcel/Documents/faucet/interface/src/lib/index.ts' is not under 'rootDir' '/Users/marcel/Documents/faucet/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by default include pattern '**/*'

error TS6059: File '/Users/marcel/Documents/faucet/interface/vite.config.ts' is not under 'rootDir' '/Users/marcel/Documents/faucet/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by default include pattern '**/*'
```

This can be fixed by removing the `rootDir` from the `tsconfig.json` of the root directory. Once done, I get
```
$ yarn build
interface/vite.config.ts:1:27 - error TS2307: Cannot find module '@sveltejs/kit/vite' or its corresponding type declarations.

1 import { sveltekit } from "@sveltejs/kit/vite";
                            ~~~~~~~~~~~~~~~~~~~~

src/server.ts:8:25 - error TS7016: Could not find a declaration file for module '../interface/build/handler.js'. '/Users/marcel/Documents/faucet/interface/build/handler.js' implicitly has an 'any' type.

8 import { handler } from "../interface/build/handler.js";
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in 2 files.

Errors  Files
     1  interface/vite.config.ts:1
     1  src/server.ts:8
```

Does the project compile for you?

**Comment by Sharqiewicz** [comment]
@ebma I added a section in README.md

```
### Develop localy

If you want to develop frontend

1. go to /interface
2. run `yarn dev`

If you want to develop both frontend and backend

1. in /interface run `yarn build`
2. in the main directory run `yarn dev`
```

If you want to run the expressJS part locally, you need to create the frontend build first.

File: README.md [hunk 4]
```diff
@@ -34,3 +34,15 @@ here defined as 8 hours.
 - `RATE_LIMIT_WINDOW_MINUTES` - Rate limit window in minutes. Defaults to `1`.
 - `RATE_LIMIT_MAX_REQUESTS` - Max requests per rate limit window. Defaults to `60`.
 - `RATE_LIMIT_NUMBER_OF_PROXIES` - Allowed number of proxies in front of the service. Defaults to `1`.
+
+### Develop localy
```
Comment (ebma) [inline README.md]:
```suggestion
### Develop locally
```

File: src/server.ts (hunk 3 above)
Comment (ebma) [reply src/server.ts]: An obvious downside of this approach is that we cannot use the backend without the frontend anymore. 

We probably wouldn't want to do this anyway, but can we maybe configure this somehow so that the express server only adds the handler if the file `handler.js` exists, and if it doesn't, we ignore it?

File: interface/src/routes/+page.svelte [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+					placeholder="2fa9..."
+					bind:value={address}
+					required
+					class="w-full px-2 py-1 text-xl text-gray-200 border rounded bg-inherit md:min-w-96"
+					id="foucoco_address"
+				/>
+				<button
+					type="submit"
+					class="w-full px-8 py-2 mt-4 font-bold text-white transition rounded-full bg-emerald-500 hover:bg-emerald-700"
+					>GET some AMPE</button
```
Comment (ebma) [inline interface/src/routes/+page.svelte]:
```suggestion
					>Get AMPE (testnet)</button
```

**Review by ebma — COMMENTED** [review]
I was able to compile it after downgrading my yarn version from 4.x to 1.22.x. 

I tested it and it works overall 🙏 Some suggestions:
- let's add the same text below the Amplitude logo, as we do in the portal, to indicate that it's not actually Amplitude but Foucoco
image


- Let's disable the 'GET some AMPE' button while the process is loading.
- Let's change the textfield to prevent autofill from password managers or similar

File: interface/src/routes/+page.svelte [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		class="absolute hidden scale-125 md:block -left-3 top-1/2"
+	/>
+	<img src="circle.svg" alt="just circle" class="absolute hidden left-1/2 md:block top-1/2" />
+	<img src="circle.svg" alt="just circle" class="absolute scale-150 right-1/2 top-2/3" />
+	<section class="flex items-center justify-center mt-10">
+		<article
+			class="relative z-50 flex flex-col items-center max-w-3xl px-20 py-10 bg-zinc-800 rounded-2xl"
+		>
+			<h1 class="text-4xl font-bold text-center text-gray-200">Foucoco Faucet</h1>
+			<p class="mt-2 text-center text-zinc-400">Get AMPE tokens for Foucoco parachain</p>
```
Comment (ebma) [inline interface/src/routes/+page.svelte]:
```suggestion
			<p class="mt-2 text-center text-zinc-400">Get AMPE tokens for the Foucoco parachain</p>
```

**Review by ebma — COMMENTED** [review]
My password managers are still trying to fill that form :(

image

and the 'Get ampe' button is not disabled when loading, which likely leads to the user clicking on it multiple times.

Besides that, it looks good to me 👍

**Comment by Sharqiewicz** [comment]
@ebma I disabled autocomplete for 1Password and LastPassword. The input is disabled when the request is pending. 🚀

**Comment by ebma** [comment]
My password manager is still showing on the latest version, but I'll merge it either way. I fixed the remaining issues with the `yarn build` command throwing an error.
